### PR TITLE
aws_ebs_volume: gp3 volume scalable throughput

### PR DESCRIPTION
**Pull request by ewbankkit**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates #16514.

Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
resource/aws_ebs_volume: Support `type` value of `gp3`. Add `throughput` attribute.
resource/aws_ebs_volume: Add `throughput` attribute.
```

Output from acceptance testing:

```console
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSEBSVolume_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSEBSVolume_ -timeout 120m
=== RUN   TestAccAWSEBSVolume_basic
=== PAUSE TestAccAWSEBSVolume_basic
=== RUN   TestAccAWSEBSVolume_updateAttachedEbsVolume
=== PAUSE TestAccAWSEBSVolume_updateAttachedEbsVolume
=== RUN   TestAccAWSEBSVolume_updateSize
=== PAUSE TestAccAWSEBSVolume_updateSize
=== RUN   TestAccAWSEBSVolume_updateType
=== PAUSE TestAccAWSEBSVolume_updateType
=== RUN   TestAccAWSEBSVolume_updateIops_Io1
=== PAUSE TestAccAWSEBSVolume_updateIops_Io1
=== RUN   TestAccAWSEBSVolume_updateIops_Io2
=== PAUSE TestAccAWSEBSVolume_updateIops_Io2
=== RUN   TestAccAWSEBSVolume_kmsKey
=== PAUSE TestAccAWSEBSVolume_kmsKey
=== RUN   TestAccAWSEBSVolume_NoIops
=== PAUSE TestAccAWSEBSVolume_NoIops
=== RUN   TestAccAWSEBSVolume_InvalidIopsForType
=== PAUSE TestAccAWSEBSVolume_InvalidIopsForType
=== RUN   TestAccAWSEBSVolume_withTags
=== PAUSE TestAccAWSEBSVolume_withTags
=== RUN   TestAccAWSEBSVolume_multiAttach
=== PAUSE TestAccAWSEBSVolume_multiAttach
=== RUN   TestAccAWSEBSVolume_outpost
=== PAUSE TestAccAWSEBSVolume_outpost
=== RUN   TestAccAWSEBSVolume_gp3_basic
=== PAUSE TestAccAWSEBSVolume_gp3_basic
=== RUN   TestAccAWSEBSVolume_gp3_iops
=== PAUSE TestAccAWSEBSVolume_gp3_iops
=== RUN   TestAccAWSEBSVolume_gp3_throughput
=== PAUSE TestAccAWSEBSVolume_gp3_throughput
=== RUN   TestAccAWSEBSVolume_disappears
=== PAUSE TestAccAWSEBSVolume_disappears
=== CONT  TestAccAWSEBSVolume_basic
=== CONT  TestAccAWSEBSVolume_withTags
=== CONT  TestAccAWSEBSVolume_disappears
=== CONT  TestAccAWSEBSVolume_gp3_throughput
=== CONT  TestAccAWSEBSVolume_gp3_iops
=== CONT  TestAccAWSEBSVolume_gp3_basic
=== CONT  TestAccAWSEBSVolume_outpost
=== CONT  TestAccAWSEBSVolume_multiAttach
=== CONT  TestAccAWSEBSVolume_updateIops_Io2
=== CONT  TestAccAWSEBSVolume_InvalidIopsForType
=== CONT  TestAccAWSEBSVolume_NoIops
=== CONT  TestAccAWSEBSVolume_kmsKey
=== CONT  TestAccAWSEBSVolume_updateType
=== CONT  TestAccAWSEBSVolume_updateIops_Io1
=== CONT  TestAccAWSEBSVolume_updateSize
=== CONT  TestAccAWSEBSVolume_updateAttachedEbsVolume
=== CONT  TestAccAWSEBSVolume_outpost
    data_source_aws_outposts_outposts_test.go:66: skipping since no Outposts found
--- SKIP: TestAccAWSEBSVolume_outpost (2.13s)
--- PASS: TestAccAWSEBSVolume_InvalidIopsForType (17.40s)
--- PASS: TestAccAWSEBSVolume_disappears (52.68s)
--- PASS: TestAccAWSEBSVolume_NoIops (55.37s)
--- PASS: TestAccAWSEBSVolume_gp3_basic (64.85s)
--- PASS: TestAccAWSEBSVolume_basic (65.87s)
--- PASS: TestAccAWSEBSVolume_multiAttach (66.08s)
--- PASS: TestAccAWSEBSVolume_withTags (66.99s)
--- PASS: TestAccAWSEBSVolume_kmsKey (67.13s)
--- PASS: TestAccAWSEBSVolume_updateIops_Io1 (95.66s)
--- PASS: TestAccAWSEBSVolume_gp3_iops (96.68s)
--- PASS: TestAccAWSEBSVolume_updateType (98.48s)
--- PASS: TestAccAWSEBSVolume_updateIops_Io2 (98.71s)
--- PASS: TestAccAWSEBSVolume_gp3_throughput (99.21s)
--- PASS: TestAccAWSEBSVolume_updateSize (99.32s)
--- PASS: TestAccAWSEBSVolume_updateAttachedEbsVolume (181.29s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	181.371s
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSEbsVolumeDataSource_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSEbsVolumeDataSource_ -timeout 120m
=== RUN   TestAccAWSEbsVolumeDataSource_basic
=== PAUSE TestAccAWSEbsVolumeDataSource_basic
=== RUN   TestAccAWSEbsVolumeDataSource_multipleFilters
=== PAUSE TestAccAWSEbsVolumeDataSource_multipleFilters
=== CONT  TestAccAWSEbsVolumeDataSource_basic
=== CONT  TestAccAWSEbsVolumeDataSource_multipleFilters
--- PASS: TestAccAWSEbsVolumeDataSource_basic (25.19s)
--- PASS: TestAccAWSEbsVolumeDataSource_multipleFilters (26.17s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	26.219s
```

### CODE REVIEW

File: aws/resource_aws_ebs_volume.go [hunk 1]
```diff
@@ -286,6 +298,7 @@ func resourceAwsEbsVolumeRead(d *schema.ResourceData, meta interface{}) error {
 	d.Set("snapshot_id", aws.StringValue(volume.SnapshotId))
 	d.Set("outpost_arn", aws.StringValue(volume.OutpostArn))
 	d.Set("multi_attach_enabled", volume.MultiAttachEnabled)
+	d.Set("throughput", aws.Int64Value(volume.Throughput))
```
Comment (mvisonneau) [inline aws/resource_aws_ebs_volume.go:301]: to avoid having to configure it when using `gp3`, should we default it to `125` if the `VolumeType === gp3`? 🤔

Comment (ewbankkit) [reply aws/resource_aws_ebs_volume.go:301]: The [API](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_CreateVolume.html#API_CreateVolume_RequestParameters) specifies no default although you are correct that if it's not specified for `gp3` volumes the `DescribeVolumes` call returns a value of `125`.
The attribute is declared as `Computed: true` to handle this.

File: aws/resource_aws_ebs_volume.go [hunk 2]
```diff
@@ -373,3 +371,40 @@ func resourceAwsEbsVolumeDelete(d *schema.ResourceData, meta interface{}) error
...
+		// Hard validation in place to return an error if IOPs are provided
+		// for an unsupported storage type.
+		// Reference: https://github.com/hashicorp/terraform-provider-aws/issues/12667
+		switch volumeType {
+		case ec2.VolumeTypeIo1, ec2.VolumeTypeIo2:
+			if iops == 0 {
+				return fmt.Errorf("'iops' must be set when 'type' is '%s'", volumeType)
+			}
+
+		case ec2.VolumeTypeGp3:
```
Comment (jayzalowitz) [inline aws/resource_aws_ebs_volume.go:394]: If throughput > 1000 it should fail if im not wrong about my assumption here.

Comment (jayzalowitz) [reply aws/resource_aws_ebs_volume.go:394]: Also pretty sure you cant go under 125

File: aws/resource_aws_ebs_volume_test.go [hunk 3]
```diff
@@ -89,6 +89,7 @@ func TestAccAWSEBSVolume_basic(t *testing.T) {
 					resource.TestCheckResourceAttr(resourceName, "type", "gp2"),
 					resource.TestCheckResourceAttr(resourceName, "outpost_arn", ""),
 					resource.TestCheckResourceAttr(resourceName, "multi_attach_enabled", "false"),
+					resource.TestCheckResourceAttr(resourceName, "throughput", "0"),
```
Comment (jayzalowitz) [inline aws/resource_aws_ebs_volume_test.go:92]: You shouldnt test this with 0, shouldnt it be 125?

File: aws/resource_aws_ebs_volume_test.go [hunk 4]
```diff
@@ -416,6 +430,156 @@ func TestAccAWSEBSVolume_outpost(t *testing.T) {
...
+					resource.TestCheckResourceAttr(resourceName, "encrypted", "false"),
+					resource.TestCheckResourceAttr(resourceName, "iops", "3000"),
+					resource.TestCheckResourceAttr(resourceName, "kms_key_id", ""),
+					resource.TestCheckResourceAttr(resourceName, "multi_attach_enabled", "false"),
+					resource.TestCheckResourceAttr(resourceName, "outpost_arn", ""),
+					resource.TestCheckResourceAttr(resourceName, "size", "10"),
+					resource.TestCheckResourceAttr(resourceName, "snapshot_id", ""),
+					resource.TestCheckResourceAttr(resourceName, "tags.%", "1"),
+					resource.TestCheckResourceAttr(resourceName, "tags.Name", rName),
+					resource.TestCheckResourceAttr(resourceName, "throughput", "600"),
```
Comment (jayzalowitz) [inline aws/resource_aws_ebs_volume_test.go:575]: Shouldnt top test be 1k? 1001 is a good failure test

File: aws/resource_aws_ebs_volume.go [hunk 5]
```diff
@@ -373,3 +371,40 @@ func resourceAwsEbsVolumeDelete(d *schema.ResourceData, meta interface{}) error
 
 	return nil
 }
+
+func resourceAwsEbsVolumeCustomizeDiff(_ context.Context, diff *schema.ResourceDiff, meta interface{}) error {
+	iops := diff.Get("iops").(int)
```
Comment (rajivshah3) [inline aws/resource_aws_ebs_volume.go:376]: Should `throughput` be validated here as well? It's only valid if the volume type is `gp3`

Comment (ewbankkit) [reply aws/resource_aws_ebs_volume.go:376]: Added plan-time check; Thanks.

**Comment by ewbankkit** [comment]
Re-ran acceptance tests:

```console
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSEBSVolume_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSEBSVolume_ -timeout 120m
=== RUN   TestAccAWSEBSVolume_basic
=== PAUSE TestAccAWSEBSVolume_basic
=== RUN   TestAccAWSEBSVolume_updateAttachedEbsVolume
=== PAUSE TestAccAWSEBSVolume_updateAttachedEbsVolume
=== RUN   TestAccAWSEBSVolume_updateSize
=== PAUSE TestAccAWSEBSVolume_updateSize
=== RUN   TestAccAWSEBSVolume_updateType
=== PAUSE TestAccAWSEBSVolume_updateType
=== RUN   TestAccAWSEBSVolume_updateIops_Io1
=== PAUSE TestAccAWSEBSVolume_updateIops_Io1
=== RUN   TestAccAWSEBSVolume_updateIops_Io2
=== PAUSE TestAccAWSEBSVolume_updateIops_Io2
=== RUN   TestAccAWSEBSVolume_kmsKey
=== PAUSE TestAccAWSEBSVolume_kmsKey
=== RUN   TestAccAWSEBSVolume_NoIops
=== PAUSE TestAccAWSEBSVolume_NoIops
=== RUN   TestAccAWSEBSVolume_InvalidIopsForType
=== PAUSE TestAccAWSEBSVolume_InvalidIopsForType
=== RUN   TestAccAWSEBSVolume_InvalidThroughputForType
=== PAUSE TestAccAWSEBSVolume_InvalidThroughputForType
=== RUN   TestAccAWSEBSVolume_withTags
=== PAUSE TestAccAWSEBSVolume_withTags
=== RUN   TestAccAWSEBSVolume_multiAttach
=== PAUSE TestAccAWSEBSVolume_multiAttach
=== RUN   TestAccAWSEBSVolume_outpost
=== PAUSE TestAccAWSEBSVolume_outpost
=== RUN   TestAccAWSEBSVolume_gp3_basic
=== PAUSE TestAccAWSEBSVolume_gp3_basic
=== RUN   TestAccAWSEBSVolume_gp3_iops
=== PAUSE TestAccAWSEBSVolume_gp3_iops
=== RUN   TestAccAWSEBSVolume_gp3_throughput
=== PAUSE TestAccAWSEBSVolume_gp3_throughput
=== RUN   TestAccAWSEBSVolume_disappears
=== PAUSE TestAccAWSEBSVolume_disappears
=== CONT  TestAccAWSEBSVolume_basic
=== CONT  TestAccAWSEBSVolume_InvalidThroughputForType
=== CONT  TestAccAWSEBSVolume_withTags
=== CONT  TestAccAWSEBSVolume_disappears
=== CONT  TestAccAWSEBSVolume_gp3_throughput
=== CONT  TestAccAWSEBSVolume_gp3_iops
=== CONT  TestAccAWSEBSVolume_gp3_basic
=== CONT  TestAccAWSEBSVolume_outpost
=== CONT  TestAccAWSEBSVolume_multiAttach
=== CONT  TestAccAWSEBSVolume_updateIops_Io2
=== CONT  TestAccAWSEBSVolume_InvalidIopsForType
=== CONT  TestAccAWSEBSVolume_NoIops
=== CONT  TestAccAWSEBSVolume_kmsKey
=== CONT  TestAccAWSEBSVolume_updateType
=== CONT  TestAccAWSEBSVolume_updateIops_Io1
=== CONT  TestAccAWSEBSVolume_updateSize
=== CONT  TestAccAWSEBSVolume_updateAttachedEbsVolume
=== CONT  TestAccAWSEBSVolume_outpost
    data_source_aws_outposts_outposts_test.go:66: skipping since no Outposts found
--- SKIP: TestAccAWSEBSVolume_outpost (2.23s)
--- PASS: TestAccAWSEBSVolume_InvalidIopsForType (20.31s)
--- PASS: TestAccAWSEBSVolume_InvalidThroughputForType (20.38s)
--- PASS: TestAccAWSEBSVolume_disappears (56.19s)
--- PASS: TestAccAWSEBSVolume_NoIops (64.44s)
--- PASS: TestAccAWSEBSVolume_gp3_basic (70.68s)
--- PASS: TestAccAWSEBSVolume_basic (70.71s)
--- PASS: TestAccAWSEBSVolume_multiAttach (70.84s)
--- PASS: TestAccAWSEBSVolume_withTags (71.61s)
--- PASS: TestAccAWSEBSVolume_kmsKey (72.73s)
--- PASS: TestAccAWSEBSVolume_updateSize (101.87s)
--- PASS: TestAccAWSEBSVolume_gp3_throughput (103.67s)
--- PASS: TestAccAWSEBSVolume_updateIops_Io2 (104.24s)
--- PASS: TestAccAWSEBSVolume_updateType (104.26s)
--- PASS: TestAccAWSEBSVolume_updateIops_Io1 (105.10s)
--- PASS: TestAccAWSEBSVolume_gp3_iops (105.12s)
--- PASS: TestAccAWSEBSVolume_updateAttachedEbsVolume (219.29s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	219.373s
```

File: aws/resource_aws_ebs_volume_test.go (hunk 3 above)
Comment (ewbankkit) [reply aws/resource_aws_ebs_volume_test.go:92]: This test uses the default `type` ( = `gp2`) and the API returns no value (translates to 0 in state) for `throughput` unless `type` is `gp3`.

File: aws/resource_aws_ebs_volume.go (hunk 2 above)
Comment (ewbankkit) [reply aws/resource_aws_ebs_volume.go:396]: That is validated via `ValidateFunc: validation.IntBetween(125, 1000),`.
The `ValidateFunc` only gets to look at the attribute it's specified on so for cross-attribute validation cases we need to code a `CustomizeDiff` function.
Both run during `terraform plan`.

File: aws/resource_aws_ebs_volume_test.go (hunk 4 above)
Comment (ewbankkit) [reply aws/resource_aws_ebs_volume_test.go:590]: Not really testing top-of-the-range here, just that the value gets updated correctly.
We are implicitly trusting `ValidateFunc: validation.IntBetween(125, 1000),`.
Thanks.

**Comment by hakman** [comment]
Hello and thanks for trying to add this fairly new attribute.
Would it be possible to add it to the review queue?
Having it added for resources and missing in others makes things a bit confusing.

File: aws/resource_aws_ebs_volume.go [hunk 6]
```diff
@@ -373,3 +371,53 @@ func resourceAwsEbsVolumeDelete(d *schema.ResourceData, meta interface{}) error
...
+
+		// MultiAttachEnabled is supported with io1 volumes only.
+		if multiAttachEnabled && volumeType != ec2.VolumeTypeIo1 {
+			return fmt.Errorf("'multi_attach_enabled' must not be set when 'type' is '%s'", volumeType)
+		}
+
+		// Throughput is valid only for gp3 volumes.
+		if throughput > 0 && volumeType != ec2.VolumeTypeGp3 {
+			return fmt.Errorf("'throughput' must not be set when 'type' is '%s'", volumeType)
+		}
```
Comment (bflad) [inline aws/resource_aws_ebs_volume.go:412]: Just noting here that this type of validation can become a maintenance burden.

**Review by bflad — APPROVED** [review]
Looks good to me, thanks, @ewbankkit 🚀 

Output from acceptance testing in AWS Commercial:

```
--- PASS: TestAccAWSEBSVolume_basic (66.62s)
--- PASS: TestAccAWSEBSVolume_disappears (56.99s)
--- PASS: TestAccAWSEBSVolume_gp3_basic (63.55s)
--- PASS: TestAccAWSEBSVolume_gp3_iops (91.99s)
--- PASS: TestAccAWSEBSVolume_gp3_throughput (90.69s)
--- PASS: TestAccAWSEBSVolume_InvalidIopsForType (9.87s)
--- PASS: TestAccAWSEBSVolume_InvalidThroughputForType (10.03s)
--- PASS: TestAccAWSEBSVolume_kmsKey (65.84s)
--- PASS: TestAccAWSEBSVolume_multiAttach (66.28s)
--- PASS: TestAccAWSEBSVolume_NoIops (63.62s)
--- PASS: TestAccAWSEBSVolume_updateAttachedEbsVolume (170.50s)
--- PASS: TestAccAWSEBSVolume_updateIops_Io1 (91.00s)
--- PASS: TestAccAWSEBSVolume_updateIops_Io2 (91.30s)
--- PASS: TestAccAWSEBSVolume_updateSize (82.58s)
--- PASS: TestAccAWSEBSVolume_updateType (92.09s)
--- PASS: TestAccAWSEBSVolume_withTags (66.49s)
--- SKIP: TestAccAWSEBSVolume_outpost (1.68s)

--- PASS: TestAccAWSEbsVolumeDataSource_basic (57.85s)
--- PASS: TestAccAWSEbsVolumeDataSource_multipleFilters (57.39s)
```

Output from acceptance testing in AWS GovCloud (US):

```
--- FAIL: TestAccAWSEBSVolume_multiAttach (12.16s) # unrelated; will ensure there is covering issue

    resource_aws_ebs_volume_test.go:398: Step 1/2 error: Error running apply:
        Error: Error creating EC2 volume: InvalidParameterValue: The specified zone does not support multi-attach-enabled volumes.
        	status code: 400, request id: 13ded026-b104-49b4-921e-f3cc20abfdc5

--- FAIL: TestAccAWSEBSVolume_updateIops_Io2 (12.52s) # unrelated; will ensure there is covering issue

    resource_aws_ebs_volume_test.go:248: Step 1/3 error: Error running apply:
        Error: Error creating EC2 volume: UnknownVolumeType: Unsupported volume type 'io2' for volume creation.
        	status code: 400, request id: 2e43e591-c913-491c-a9dd-4bf69031059c

--- PASS: TestAccAWSEBSVolume_basic (37.24s)
--- PASS: TestAccAWSEBSVolume_disappears (31.88s)
--- PASS: TestAccAWSEBSVolume_gp3_basic (38.94s)
--- PASS: TestAccAWSEBSVolume_gp3_iops (64.65s)
--- PASS: TestAccAWSEBSVolume_gp3_throughput (64.05s)
--- PASS: TestAccAWSEBSVolume_InvalidIopsForType (5.91s)
--- PASS: TestAccAWSEBSVolume_InvalidThroughputForType (7.92s)
--- PASS: TestAccAWSEBSVolume_kmsKey (41.00s)
--- PASS: TestAccAWSEBSVolume_NoIops (35.33s)
--- PASS: TestAccAWSEBSVolume_updateAttachedEbsVolume (172.84s)
--- PASS: TestAccAWSEBSVolume_updateIops_Io1 (64.67s)
--- PASS: TestAccAWSEBSVolume_updateSize (65.03s)
--- PASS: TestAccAWSEBSVolume_updateType (64.96s)
--- PASS: TestAccAWSEBSVolume_withTags (38.69s)
--- SKIP: TestAccAWSEBSVolume_outpost (2.43s)

--- PASS: TestAccAWSEbsVolumeDataSource_basic (29.45s)
--- PASS: TestAccAWSEbsVolumeDataSource_multipleFilters (29.91s)
```

File: aws/resource_aws_ebs_volume.go (hunk 6 above)
Comment (ewbankkit) [reply aws/resource_aws_ebs_volume.go:412]: I tend to agree but it's also no fun for practitioners to be hit with an error during `apply` that could have caught at `plan` or best `validate`.

Comment (bflad) [reply aws/resource_aws_ebs_volume.go:412]: Of course, no one is arguing against that. The challenge here is that this type of information is not programmatically encoded in the API model and can differ per partition (or in worse cases per region within a partition), which means it needs manual curation should changes need to occur over time. It is no fun for practitioners to also be artificially limited in the future from doing a perfectly valid API call while waiting for this project to be updated. The most pragmatic answer probably lies somewhere in between both ends of the spectrum, adding it where there is a strong community ask and a strong maintainer ability to keep it up to date versus omitting it when either is not true.

**Pull request merged**